src/names.c

```diff
@@ -0,0 +1,379 @@
+#include "vctrs.h"
+#include "dictionary.h"
+#include "utils.h"
+
+static void describe_repair(SEXP old, SEXP new);
+
+
+// [[ register() ]]
+SEXP vec_bare_names(SEXP x) {
+  if (OBJECT(x) && Rf_inherits(x, "data.frame")) {
+    return R_NilValue;
+  }
+
+  if (vec_dim(x) == 1) {
+    if (OBJECT(x)) {
+      return vctrs_dispatch1(syms_names, fns_names, syms_x, x);
+    } else {
+      return r_names(x);
+    }
+  }
+
+  SEXP dimnames = PROTECT(Rf_getAttrib(x, R_DimNamesSymbol));
+  if (dimnames == R_NilValue || Rf_length(dimnames) < 1) {
+    UNPROTECT(1);
+    return R_NilValue;
+  }
+
+  SEXP out = VECTOR_ELT(dimnames, 0);
+  UNPROTECT(1);
+  return out;
+}
+
+static struct sexp_cow as_minimal_names(struct sexp_cow cow_names) {
```

Comment:
I thought we should avoid allocs and copies if we can, this will be called once per inputs, which can be many when assembling data frames from vectors with e.g. purrr.

The structure is always the same: first pass to check if a modif is needed, and then second pass with the SEXP copied.

Reply:
Sorry, on second inspection the cow structure was completely silly. I have removed it. Is this now better?